org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestManager.java

```diff
@@ -51,18 +55,26 @@ public void shutdown() {
 		parallel.shutdown();
 		cancel();
 	}
-
+	
+	protected final OperationCanceledManager getOperationCanceledManager() {
+		return operationCanceledManager;
+	}
+	
+	protected final ExecutorService getParallelExecutorService() {
+		return parallel;
+	}
+	
 	/**
 	 * Run the given cancellable logic as a read request.
 	 */
-	public <V> CompletableFuture<V> runRead(Function1<? super CancelIndicator, ? extends V> cancellable) {
+	public synchronized <V> CompletableFuture<V> runRead(Function1<? super CancelIndicator, ? extends V> cancellable) {
```

Comment:
Is `synchronized` necessary here?  IIUIC then there is a single connection (socket) that is feeding us read and write commands.  I think therefore there is only one thread invoking `runRead` and `runWrite` in a nice, sequential way?

Reply:
Client code (especially ExecuteCommandServices) can inject the RequestManager to trigger write requests from arbitrary threads. That's why we do need the sync here.